pbft/pbft.cpp

```diff
@@ -650,24 +722,69 @@ pbft::unstable_checkpoints_count() const
 void
 pbft::maybe_stabilize_checkpoint(const checkpoint_t& cp)
 {
-    if (this->local_unstable_checkpoints.count(cp) == 0 || this->unstable_checkpoint_proofs[cp].size() < this->quorum_size())
+    if (this->unstable_checkpoint_proofs[cp].size() < this->quorum_size())
     {
         return;
     }
 
+    if (this->local_unstable_checkpoints.count(cp) != 0)
+    {
+        this->stabilize_checkpoint(cp);
+    }
+    else
+    {
+        // we don't have this checkpoint, so we need to catch up
+        this->request_checkpoint_state(cp);
+    }
+}
+
+void
+pbft::stabilize_checkpoint(const checkpoint_t& cp)
+{
     this->stable_checkpoint = cp;
     this->stable_checkpoint_proof = this->unstable_checkpoint_proofs[cp];
 
     LOG(info) << boost::format("Checkpoint %1% at seq %2% is now stable; clearing old data")
-              % cp.second
-              % cp.first;
+        % cp.second % cp.first;
 
     this->clear_local_checkpoints_until(cp);
     this->clear_checkpoint_messages_until(cp);
     this->clear_operations_until(cp);
 
     this->low_water_mark = std::max(this->low_water_mark, cp.first);
-    this->high_water_mark = std::max(this->high_water_mark, cp.first + std::lround(HIGH_WATER_INTERVAL_IN_CHECKPOINTS*CHECKPOINT_INTERVAL));
+    this->high_water_mark = std::max(this->high_water_mark,
+        cp.first + std::lround(HIGH_WATER_INTERVAL_IN_CHECKPOINTS * CHECKPOINT_INTERVAL));
+
+}
+
+void
+pbft::request_checkpoint_state(const checkpoint_t& cp) const
+{
+    pbft_membership_msg msg;
+    msg.set_type(PBFT_MMSG_GET_STATE);
+    msg.set_sequence(cp.first);
+    msg.set_state_hash(cp.second);
+
+    LOG(info) << boost::format("Requesting checkpoint state for hash %1% at seq %2%")
+        % cp.second % cp.first;
+
+    auto msg_ptr = std::make_shared<bzn::encoded_message>(this->wrap_message(msg));
+    this->node->send_message_str(make_endpoint(this->get_primary()), msg_ptr);
```

Comment:
I'd rather not ask the primary. Instead:
- look at the checkpoint proof to determine who allegedly has the state (primary may not even have it)
- send the request to one of those nodes at random to spread load

Reply:
done